rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,337 @@
+# Composable Pipelines
+
+- **Status**: In-Review
+- **Created**: Aug 19, 2022
+- **ETA**: Sep 19, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Contributors**:
+  - [Anthony Verez](https://github.com/netantho)
+  - [Dominik Lohmann](https://github.com/dominiklohmann)
+  - [Rémi Dettai](https://github.com/rdettai)
+- **Discussion**: [PR #2511](https://github.com/tenzir/vast/pull/2511)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      summarize count(dst) group-by src |
+      write feather |
+      to /path/to/file.feather'
+```
+
+Just to illustrate the composability, we could take this to an extreme when
+performance doesn't matter:
+
+```bash
+vast from s3://aws |
+  vast read json |
+  vast 'summarize count(dst) group-by src' |
+  vast write feather |
+  vast to /path/to/file.feather
+```
+
+The UX would translate seamlessly to other languages, e.g., Python:
+
+```python
+import vast as v
+await v.from("s3://aws")
+       .read_json() # or read("json")
+       .group_by("src") # could also be merged with summarize()
+       .summarize(v.count("dst"))
+       .write_feather() # or write("feather")
+       .to("/path/to/file.feather")
+```
```

Comment:
We don't need the "extreme" case above to build this in Python.

Reply:
I don't understand the action item for this discussion point. Was it just sharing an opinion?

Reply:
I didn't read it that way, but agreed that it can be confusing. Maybe putting the "extreme" case in a separate "Remark" section would help the reader.

Reply:
To clarify: The quoted paragraph is written as if it depended on the "extreme" case shown directly above, and that I don't believe to be true. They should be presented independently.

Reply:
Got it. I moved this to the bottom of the proposal.